algorithms/a3c/env_vizdoom.py

```diff
@@ -0,0 +1,379 @@
+        """Executes an action in the environment to reach a new state.
+        Args:
+          action_id: An integer corresponding to the action.
+        Returns:
+           state: A tuple of screen buffer state and instruction.
+           reward: Reward at that step.
+           is_final: Flag indicating terminal state.
+           extra_args: Dictionary of additional arguments/parameters.
+        """
+        # Repeat the action for 5 frames.
+        if self.params.visualize:
+            # Render 5 frames for better visualization.
+            for _ in range(5):
+                self.game.make_action(actions[action_id], 1)
+                # Slowing down the game for better visualization.
+                sleep(self.params.sleep)
+        else:
+            self.game.make_action(actions[action_id], 5)
+
+        self.time += 1
+        reward = self.get_reward()
+
+        # End the episode if episode limit is reached or
+        # agent reached an object.
+        is_final = True if self.time == self.params.max_episode_length \
+            or reward != self.params.living_reward else False
+
+        screen = self.game.get_state().screen_buffer
+        screen_buf = process_screen(
+            screen, self.params.frame_height, self.params.frame_width)
+
+        state = (screen_buf, self.instruction)
+
+        return state, reward, is_final, None
+
+    def close(self):
+        self.game.close()
+
+    def get_reward(self):
+        """Get the reward received by the agent in the last time step."""
+        # If agent reached the correct object, reward = +1.
+        # If agent reach a wrong object, reward = -0.2.
+        # Otherwise, reward = living_reward.
+        self.agent_x, self.agent_y = get_agent_location(self.game)
+        target_location = self.object_coordinates[self.correct_location]
+        dist = get_l2_distance(self.agent_x, self.agent_y,
+                               target_location.x, target_location.y)
+        if dist <= REWARD_THRESHOLD_DISTANCE:
+            reward = CORRECT_OBJECT_REWARD
+            return reward
+        else:
+            for i, object_location in enumerate(self.object_coordinates):
+                if i == self.correct_location:
+                    continue
+                dist = get_l2_distance(self.agent_x, self.agent_y,
+                                       object_location.x, object_location.y)
+                if dist <= REWARD_THRESHOLD_DISTANCE:
+                    reward = WRONG_OBJECT_REWARD
+                    return reward
+            reward = self.params.living_reward
+
+        return reward
+
+    def get_agent_and_object_positions(self):
+        """Get agent and object positions based on the difficulty
+        of the environment.
+        """
+        object_x_coordinates = []
+        object_y_coordinates = []
+
+        if self.params.difficulty == 'easy':
+            # Agent location fixed in Easy.
+            agent_x_coordinate = 128
```

Comment:
If you love so much vertical space, you can put agent x, y and orientation in one line
```suggestion
            agent_x_coordinate, agent_y_coordinate, agent_orientation = 128, 512, 0
```

Reply:
since agent_x, y and rotation are the same for easy and medium I would write it outside the if elses to avoid repeating code 
```suggestion
            agent_x_coordinate = 128
    # Then this
          if self.params.difficulty == 'easy':
               # only code to generate x and y.
          if self.params.difficulty == 'medium':
               # only code to generate x and y.
          if self.params.difficulty == 'hard':
               # override x, y and orientation defaults
```
Also one thing, is it possible to have difficulty easy and medium? If not use if elif else instead of several ifs

Reply:
Ok I'll put in one line 

Explicit is good sometimes and in the future it might be good to change it specifically for medium. But fine i'll initialise it outside.

But im really agonizing about being careful about changing this code without testing it. Sometimes style and other changes can break it no matter how much you reason about the output. This is why we shouldn't really change much of this code. BUT FINE